### DISCUSSION TITLE
Data Loaders and Field Resolvers

### DISCUSSION THREAD

**Discussion by rpf3** [ChilliCream/hotchocolate#5615] [Q&A]
I'm looking for guidance on the recommended way to implement resolvers with the following data model:

``` c#
public class User
{
    public int Id { get; set; }
    public string Name { get; set; }
}

public class Comment
{
    public int Id { get; set; }
    public string Text { get; set; }
    public int UserId { get; set; }

    public async Task<User> GetUser(UserDataLoader userDataLoader)
   {
        // await LoadAsync(UserId);
    }
}

public class Post
{
    public int Id { get; set; }
    public string Text { get; set; }
    public int UserId { get; set; }
    public ICollection<int> CommentIds { get; set; }

    public async Task<User> GetUser(UserDataLoader userDataLoader)
   {
        // await LoadAsync(UserId);
    }

    public async Task<Comment[]> GetComments(CommentDataLoader commentDataLoader)
   {
        // await LoadAsync(CommentIds);
    }
}
```

When I query for posts, I see two database queries to the User table, the first for the Post and the second for all the comments.  I thought I understood that the data loader was supposed to bundle these together.  But as I thought about it, I realized that it would have no way of knowing to wait for the response of the comment data loader.  So I am thinking that I am missing something but I've been scouring the documentation and example workshop repositories and am unable to see where I am going wrong here.

**Comment by PascalSenn** [comment]
@rpf3  This is correct.

The execution engine batches the DataLoader calls and executes the DataLoader when it does not execute anything else. Meaning, we execute as long as we can, when we have nothing to do, we dispatch the dataloaders. 

if in your case the comments were not dataloaded, but rather on the Post itself with a sync resolver/property. It's likely that also these would make it into the first batch to the users table.

The dataloader also makes sure that the user with id `1` will be the same user throughout the whole graph. This means, if you load a comment that user `1` has written, and this user is already loaded, we just return the value from the cache without a call to the database.
